File: webapp/src/main/scala/com/graphbrain/webapp/VisualGraph.scala
package com.graphbrain.webapp

import com.graphbrain.gbdb.UserOps
import com.graphbrain.gbdb.Vertex
import com.graphbrain.gbdb.Edge
import com.graphbrain.gbdb.TextNode
import com.graphbrain.gbdb.URLNode
import com.graphbrain.gbdb.UserNode
import com.graphbrain.gbdb.ContextNode
import com.graphbrain.gbdb.ID
import com.graphbrain.gbdb.IdFamily
import com.graphbrain.gbdb.JSONGen

import com.graphbrain.gbdb.VertexStore
import com.graphbrain.gbdb.UserManagement
import com.graphbrain.gbdb.SimpleCaching
import scala.math


object VisualGraph {
  val MAX_SNODES = 15


  def generate(rootId: String, store: UserOps, user: UserNode, edgeType: String = "", relPos: Integer = -1) = {
    val userId = if (user != null) user.id else ""
    
    // get neighboring edges
    val hyperEdges = store.neighborEdges2(rootId, userId, edgeType, relPos)
    
    // map hyperedges to visual edges
    val visualEdges = hyperEdges.map(hyper2edge(_, rootId)).filter(_ != null)

    // group nodes by edge type
    val edgeNodeMap = generateEdgeNodeMap(visualEdges, rootId)

    // truncate edge node map
    val truncatedEdgeNodeMap = truncateEdgeNodeMap(edgeNodeMap, MAX_SNODES)

    // full relations list
    val allRelations = edgeNodeMap.keys.map(x => Map(("rel" -> x._1), ("pos" -> x._2),
      ("label" -> linkLabel(x._1)), ("snode" -> snodeId(x._1, x._2))))

    // create map with all information for supernodes
    val snodeMap = generateSnodeMap(truncatedEdgeNodeMap, store, rootId, user)

    // contexts
    val contexts = if ((user != null) && (user.contexts != null)) {
      user.contexts.map(c => Map(("id" -> c.id), ("name" -> c.name), ("access" -> c.access)))
    }
    else {
      Nil
    }

    // create reply structure with all the information needed for rendering
    val reply = Map(("user" -> userId), ("root" -> node2map(rootId, "", store, rootId, user)), 
      ("snodes" -> snodeMap), ("allrelations" -> allRelations), ("context" -> ID.contextId(rootId)), ("contexts" -> contexts))

    Server.store.clear()

    // generate json reply
    JSONGen.json(reply)
  }

  private def hyper2edge(edge: Edge, rootId: String) = {
    if (edge.participantIds.length > 2) {
      if (edge.edgeType == "rtype/1/instance_of~owned_by") {
        if (edge.participantIds(0) == rootId) {
          Edge("rtype/1/has", List(edge.participantIds(2), rootId), edge)
        }
        else if (edge.participantIds(2) == rootId) {
          Edge("rtype/1/has", List(rootId, edge.participantIds(0)), edge)
        }
        else {
          null
        }
      }
      if (edge.edgeType == "rtype/1/has~of_type") {
        val edgeType = "has " + ID.lastPart(edge.participantIds(2))
        Edge(edgeType, List(edge.participantIds(0), edge.participantIds(1)), edge)
      }
      else {
        val parts = edge.edgeType.split("~").toList
        val edgeType = parts.head + " " + ID.lastPart(edge.participantIds(1)) + " " + parts.tail.reduceLeft(_ + " " + _)
        //val edgeType = edge.edgeType.replaceAll("~", " .. ") + " .. "
        Edge(edgeType, List(edge.participantIds(0), edge.participantIds(2)), edge)
      }
    }
    else {
      edge
    }
  }

  private def generateEdgeNodeMap(edges: Set[Edge], rootId: String) = {
    edges.map(
      e => e.participantIds
        .zip(0 until e.participantIds.length)
        .map(x => (e.edgeType, x._2, x._1, e.getOriginalEdge.toString))
    ).flatten
      .filter(x => x._3 != rootId)
      .groupBy(x => (x._1, x._2))
      .mapValues(x => x.map(y => (y._3, y._4)))
  }

  private def truncateEdgeNodeMap(edgeNodeMap: Map[(String, Int), Set[(String, String)]], maxSize: Int) = {
    edgeNodeMap.zipWithIndex.filter(x=> x._2 < maxSize).map(x => x._1)
  }

  private def node2map(nodeId: String, nodeEdge: String, store: UserOps, rootId: String, user: UserNode) = {
    val family = IdFamily.family(nodeId)
    val node = if ((nodeId != rootId) && ((family == IdFamily.Global) || (family == IdFamily.UserSpace))) {
      TextNode.fromId(nodeId, store)
    }
    else{
      try {
        store.get(nodeId)
      }
      catch {
        case _ => null
      }
    }

    node match {
      case tn: TextNode => Map(("id" -> tn.id), ("type" -> "text"), ("text" -> tn.text), ("edge" -> nodeEdge))
      case un: URLNode => {
        val title = if (un.title == "") un.url else un.title
        Map(("id" -> un.id), ("type" -> "url"), ("text" -> title), ("url" -> un.url), ("icon" -> un.icon), ("edge" -> nodeEdge))
      }
      case un: UserNode => Map(("id" -> un.id), ("type" -> "user"), ("text" -> un.name), ("edge" -> nodeEdge))
      case cn: ContextNode => Map(("id" -> cn.id), ("type" -> "context"), ("text" -> user.name), ("text2" -> ID.humanReadable(cn.id)), ("edge" -> nodeEdge))
      case null => ""
      case _ => Map(("id" -> node.id), ("type" -> "text"), ("text" -> node.id), ("edge" -> nodeEdge))
    }
  }

  private def snodeId(edgeType: String, pos: Integer) = edgeType.replaceAll("/", "_").replaceAll(" ", "_").replaceAll("\\.", "_") + "_" + pos

  private def generateSnode(pair: ((String, Int), Set[(String, String)]), store: UserOps, rootId: String, user: UserNode) = {
    val id = snodeId(pair._1._1, pair._1._2)
    val label = linkLabel(pair._1._1)
    val color = linkColor(label)
    val nodes = pair._2.map(x => node2map(x._1, x._2, store, rootId, user))

    val data = Map(("nodes" -> nodes), ("etype" -> pair._1._1), ("rpos" -> pair._1._2), ("label" -> label), ("color" -> color))

    id -> data
  }

  private def generateSnodeMap(edgeNodeMap: Map[(String, Int), Set[(String, String)]], store: UserOps, rootId: String, user: UserNode) = {
    edgeNodeMap.map(x => generateSnode(x, store, rootId, user))
  }

  private def linkColor(label: String) = {
    val index = math.abs(label.hashCode) % Colors.colors.length
    Colors.colors(index)
  }

  private def fixLabel(label: String) = if (EdgeLabelTable.table.contains(label)) EdgeLabelTable.table(label) else label

  private def linkLabel(edgeType: String): String = {
    if (edgeType == "")
      return ""
    val lastPart = ID.parts(edgeType).last
    fixLabel(lastPart.replace("_", " "))
  }

  /*
  def main(args: Array[String]) {
    val store = new VertexStore with SimpleCaching with UserOps with UserManagement
      
    println(VisualGraph.generate("1/eraserhead", store, null))

    sys.exit(0)
  }*/
}

File: src/main/clojure/graphbrain/web/visualgraph.clj
(ns graphbrain.web.visualgraph
  (:require [clojure.data.json :as json]
            [graphbrain.web.common :as common]
            [graphbrain.web.colors :as colors]
            [graphbrain.db.gbdb :as gb]
            [graphbrain.db.maps :as maps]
            [graphbrain.db.id :as id]
            [graphbrain.db.edgetype :as edgetype]
            [graphbrain.db.entity :as entity]
            [graphbrain.db.vertex :as vertex]
            [graphbrain.db.urlnode :as url]
            [graphbrain.db.text :as text]
            [graphbrain.db.context :as context])
  (:import (com.graphbrain.web EdgeLabelTable)))

(def ^:const max-snodes 15)

(defn- hide?
  [edge]
  (let [et (maps/edge-type edge)]
    (or (= et "r/+title"))))

(defn- hide-edges
  [edges]
  (filter #(not (hide? %)) edges))

(defn- snode-id
  [rel-pos]
  (str (-> (first rel-pos)
           (clojure.string/replace #"\W" "_"))
       "_" (second rel-pos)))

(defn- hyper->edge
  [edge root-id]
  (let [c (count (maps/participant-ids edge))]
    (cond
     (> c 2)
     (let [parts (.split (maps/edge-type edge) "~")]
       {:edge-type (str (first parts)
                        " "
                        (id/last-part (nth (maps/participant-ids edge) 1))
                        " "
                        (clojure.string/join " " (rest parts)))
        :id1 (nth (maps/participant-ids edge) 0)
        :id2 (nth (maps/participant-ids edge) 2)
        :score (:score edge)
        :parent edge})
     (= c 2)
     {:edge-type (nth (maps/ids edge) 0)
      :id1 (nth (maps/ids edge) 1)
      :id2 (nth (maps/ids edge) 2)
      :score (:score edge)
      :parent edge}
     :else nil)))

(defn- add-to-edge-node-map
  [en-map key e root-id]
  (if (or (and (= (second key) 0) (= (:id1 e) root-id))
          (and (= (second key) 1) (= (:id2 e) root-id)))
    en-map
    (assoc en-map key (conj (en-map key) e))))

(defn- add-edge-to-edge-node-map
  [en-map edge root-id]
  (let [enm (add-to-edge-node-map en-map [(:edge-type edge) 0] edge root-id)
        enm (add-to-edge-node-map enm [(:edge-type edge) 1] edge root-id)]
    enm))

(defn- edge-node-map
  [edges root-id]
  (reduce #(add-edge-to-edge-node-map %1 %2 root-id) {} edges))

(defn- link-color
  [label]
  (nth colors/colors (mod (Math/abs (.hashCode label)) (count colors/colors))))

(defn- fix-label
  [label]
  (if (.containsKey EdgeLabelTable/lt label)
    (.get EdgeLabelTable/lt label)
    label))

(defn- node-label
  [node]
  (case (:type node)
    "entity" (:text node)
    "user" (:text node)
    (:type node)))

(defn- link-label
  [edge-type rpos root-node]
  (if (empty? edge-type)
    ""
    (let [rel-label (fix-label (edgetype/label edge-type))
          text (node-label root-node)]
      (if (= rpos 0)
        (str rel-label " " text)
        (str text " " rel-label)))))

(defn- node->map
  [gbdb node-id edge root-id ctxts]
  (let [vtype (id/id->type node-id)
        node (if (= vtype :entity)
               (maps/id->vertex node-id)
               (gb/getv gbdb node-id ctxts))
        node (if (nil? node) (maps/id->vertex node-id) node)
        node (maps/local->global node)
        node-edge (:id (:parent edge))
        score (:score edge)]
    (condp = (:type node)
      :entity (let [sub (entity/subentities node)
                    sub (map #(hash-map :id (id/eid->id %)
                                        :text (entity/description %)) sub)]
                {:id (:id node)
                 :type "entity"
                 :text (entity/label node-id)
                 :sub sub
                 :edge node-edge
                 :score score})
      :url (let [title (url/title gbdb (:id node) ctxts)
                 url (url/url node)]
             {:id (:id node)
              :type "url"
              :text (if (empty? title) (url/url node) title)
              :url url
              :icon (str "http://g.etfv.co/" url)
              :edge node-edge
              :score score})
      :user {:id (:id node)
             :type "user"
             :text (:name node)
             :sub [{:id "" :text "GraphBrain user"}]
             :edge node-edge
             :score score}
      :text {:id (:id node)
             :type "text"
             :text (:text (text/id->text gbdb (:id node)))
             :edge node-edge
             :score score}
      {:id (:id node)
       :type "text"
       :text (:id node)
       :edge node-edge
       :score score})))

(defn- se->node-id
  [se rp]
  (if (= (second rp) 0) (:id1 se) (:id2 se)))

(defn- snode
  [gbdb rp sedges root-node ctxts]
  (let [etype (first rp)
        rpos (second rp)
        label (link-label etype rpos root-node)
        color (link-color etype)
        nodes (map #(node->map gbdb
                               (se->node-id % rp)
                               %
                               (:id root-node)
                               ctxts) sedges)]
    {:nodes nodes
     :etype etype
     :rpos rpos
     :label label
     :color color}))

(defn- snode-map
  [gbdb en-map root-node ctxts]
  (loop [enm en-map
         count 0
         snode-map {}]
    (if (or (> count max-snodes) (empty? enm)) snode-map
        (let [en (first enm)
              rp (first en)
              snid (snode-id rp)]
          (recur (rest enm) (inc count)
                 (assoc snode-map
                   snid
                   (snode gbdb rp (second en) root-node ctxts)))))))

(defn- snode-limit-size
  [snode]
  (let [max-nodes 10
        nodes (:nodes snode)]
    (if (> (count nodes) max-nodes)
      (let [nodes (sort-by :score nodes)
            nodes (take-last max-nodes nodes)]
        (assoc snode :nodes nodes))
      snode)))

(defn- snodes-limit-size
  [snodes]
  (into {} (for [[k v] snodes]
             [k (snode-limit-size v)])))

(defn generate
  [gbdb root-id user ctxts all-ctxts]
  (let [user-id (if user (:id user) "")
        root-id (gb/id->eid gbdb root-id)
        root-id (id/local->global root-id)
        hyper-edges (gb/id->edges gbdb root-id ctxts)
        hyper-edges (hide-edges hyper-edges)
        visual-edges (filter (complement nil?)
                             (map #(hyper->edge % root-id)
                                  (filter maps/positive? hyper-edges)))
        edge-node-map (edge-node-map visual-edges root-id)
        root-node (node->map gbdb root-id "" root-id ctxts)
        all-relations (into [] (for [[rp v] edge-node-map]
                                 {:rel (first rp)
                                  :pos (second rp)
                                  :label (link-label
                                          (first rp)
                                          (second rp)
                                          root-node)
                                  :snode (snode-id rp)}))
        snode-map (snode-map gbdb edge-node-map root-node ctxts)
        snode-map (assoc snode-map :root
                         {:nodes [root-node]})
        snode-map (snodes-limit-size snode-map)]
    (json/write-str {:user user-id
                     :snodes snode-map
                     :allrelations all-relations
                     :ctxts (map #(hash-map :id % :name (context/label %)) all-ctxts)})))
